### PR TITLE
A Vanilla Node.js REST API Without Frameworks such as Express.js

### LINKED ISSUE #2029: A Vanilla Node.js REST API Without Frameworks such as Express.js
We ask that you please be patient as our team works through approving and publishing all articles/tutorials in a timely manner. Allow 1-3 days for a topic to be reviewed and/or approved - allow 3-7 days for an articles to be reviewed and/or published.

Be sure to visit our [Resources Page](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md) for tools, resources, and example articles to go over.

**NOTE: (Please ensure that you have only one open issue + linked pull request at a time. This will ensure that we complete the article in a timely manner from inception to publishing.)**

**NOTE: We tend to stray away or tend not to publish reviews/comparisons of commercial product offerings.**

*We're looking for the first 2-3 paragraphs of the article that appropriately summarize what your article will be about.*

### Proposed title of article
Your title should be descriptive of the article/tutorial. Be Specific. *Use keyword research to gain your article higher ranking.*

### Introduction paragraph (2-3 paragraphs):
Node.js is a popular very JavaScript framework. Node.js has been used to create applications of all levels. It shines most when used to build back-end services and APIs.

In most cases, when building APIs with Node.js, developers use open-source [Node.js libraries and frameworks](/engineering-education/most-useful-nodejs-packages/) such as [Express.js](/engineering-education/express/). These libraries are readily available in the NPM registry.

Whenever you use such packages, there is a lot of abstraction as you don't utilize the core functionalities of Node.js. The low-level logic Node.js of your application is hidden from you as these packages process and execute raw Node.js behind the scenes.

One key thing to note is that these packages make [Node.js a popular technology](/engineering-education/why-node-js-is-popular/). On the other side, you can opt to use core Node.js to develop your applications. This way, you make use of the [vanilla Node.js](/engineering-education/pure-node-js-no-frameworks-or-packages/) functionalities. This blog will teach you how to use vanilla Node.js with no frameworks to build simple APIs.

### Goal
In this guide, we'll build a simple REST API skeleton using functionalities core to Node.js itself. We're just using the barebones of [vanilla Node.js](https://nodejs.dev/learn) with the [HTTP module](https://nodejs.org/api/http.html#http_http) for creating and managing a server. This means we won't use NPM. Therefore no NPM associated dependencies, no `package.json`,  no `package-lock.json`, and no `node_module` folder. The reason is to demonstrate and give you an overview of how raw Node.js works/can be used without frameworks.

> Note: When working on any actual project, it is best to use Node.js libraries and packages. This way, you'll take full advantage of the readily available code to make your development workflow easier and faster.

### Key takeaways: 
Setting up a simple HTTP server
Setting up the REST API
Explore the API using Postman

### References: 
https://nodejs.dev/
https://nodejs.org/api/http.html#http_http

### Templates to use as guides
- [How To Guide Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/how-to-guide-template.md)
- [Software Review Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/software-review-template.md)
- [Tutorial Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/tutorial-template.md)

### PR DESCRIPTION
**Pull request by Rose-stack**
### Article checklist
**NOTE: (Please ensure that you have only one open issue + linked pull request at a time. This will ensure that we complete the article in a timely manner from inception to publishing.)**

**If you have not already please go over our [Resources Page](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md) for more tips and tricks on how to improve our overall technical writing.**

#### Preliminary Checks - Formatting and Structure 
- [x] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**
- [x] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [x] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**
- [x] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [x] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [x] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [x] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**
- [x] Have you included a hero image that is 600x400 pixels and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
- [x] In your PR - add a comment that states **["This fixes #(enter your Idea Content Form #)"](https://youtu.be/TKJ4RdhyB5Y?t=278)** to link and close your Idea Content (issue) when the PR is merged.
      
#### Grammar & plagiarism checks
- [x] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) -  [Grammarly](https://grammarly.com) 
- [x] Please place your article through a 3rd party plagiarism checker? We suggest using Quetext, [this tool](https://www.quetext.com/) is free and has a daily limit. **We typically accept articles with 10% or less.**
- [x] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [x] Have you added sources for quotes and images that aren't yours?

#### Technical checks
- [x] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**

### Contribution guidelines
For first-time contributors and for more details, **see [Contributing Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**
fixes #2029

### CODE REVIEW

**Comment by Rose-stack** [comment]
Fixes #2029

File: articles/a-raw-nodejs-rest-api-without-frameworks-such-as-expressjs/index.md [hunk 1]
```diff
@@ -0,0 +1,451 @@
...
+
+```js
+const PORT = process.env.PORT || 5000;
+```
+
+3. To create the server, you just need the HTTP module and then call the `createServer` method. I.e, `http.createServer`. Pass a response and a request that serves up your information. Then use;
+
+- `req.url` to set the request access route/URL.
+- `req.method` 
+- `res.writeHead` to  set any response headers.
```
Comment (kayere) [inline articles/a-raw-nodejs-rest-api-without-frameworks-such-as-expressjs/index.md:47]: Remove extra space

File: articles/a-raw-nodejs-rest-api-without-frameworks-such-as-expressjs/index.md [hunk 2]
```diff
@@ -0,0 +1,451 @@
...
+- `res.write()` to send the actual content for the.response.
```
Comment (kayere) [inline articles/a-raw-nodejs-rest-api-without-frameworks-such-as-expressjs/index.md:48]: Fix the typo at the end of the sentence.

File: articles/a-raw-nodejs-rest-api-without-frameworks-such-as-expressjs/index.md [hunk 3]
```diff
@@ -0,0 +1,451 @@
...
+- `res.end()`  to end the response.
```
Comment (kayere) [inline articles/a-raw-nodejs-rest-api-without-frameworks-such-as-expressjs/index.md:49]: Remove the extra space.

File: articles/a-raw-nodejs-rest-api-without-frameworks-such-as-expressjs/index.md [hunk 4]
```diff
@@ -0,0 +1,451 @@
...
+
+// If no route present
+  else {
+    res.writeHead(404, { "Content-Type": "application/json" });
+    res.end(JSON.stringify({ message: "Route not found" }));
+  }
+});
+```
+
+4. Call the `listen()` method and pass in the `PORT` variable. The add a `console.log()` message that will indicate the server is up and running.
```
Comment (kayere) [inline articles/a-raw-nodejs-rest-api-without-frameworks-such-as-expressjs/index.md:72]: Fix the typo at the beginning of the second sentence.

File: articles/a-raw-nodejs-rest-api-without-frameworks-such-as-expressjs/index.md [hunk 5]
```diff
@@ -0,0 +1,451 @@
...
+![Get all todos](/engineering-education/a-raw-nodejs-rest-api-without-frameworks-such-as-expressjs/postman-get-all-todos.jpg)
+
+- The response should resemble the following:
+
+![Get all todos](/engineering-education/a-raw-nodejs-rest-api-without-frameworks-such-as-expressjs/get-todos.png)
+
+#### Get a todo by id
+**/API/todos/:id GET**: This will fetch a single todo. To test it:
+
+- Open a separate tab in Postman and **SEND** a `GET` request to `http://localhost:5000/api/todos/:id`. Replace `:id` with and id of the todo you want to fetch.
```
Comment (kayere) [inline articles/a-raw-nodejs-rest-api-without-frameworks-such-as-expressjs/index.md:392]:
```suggestion
- Open a separate tab in Postman and **SEND** a `GET` request to `http://localhost:5000/api/todos/:id`. Replace `:id` with an id of the todo you want to fetch.
```

File: articles/a-raw-nodejs-rest-api-without-frameworks-such-as-expressjs/index.md [hunk 6]
```diff
@@ -0,0 +1,451 @@
...
+- Open a separate tab in Postman and **SEND** a `DELETE` request to `http://localhost:5000/api/todos/:id`. Where `:id` represents the todo you want to delete.
+
+![Delete a todo](/engineering-education/a-raw-nodejs-rest-api-without-frameworks-such-as-expressjs/postman-delete-a-todo.jpg)
+
+- The response should resemble the following:
+
+![Delete a todo](/engineering-education/a-raw-nodejs-rest-api-without-frameworks-such-as-expressjs/del-todo.png)
+
+#### UPDATE a todo
+**/API/todos/:id PATCH**: This will update a to indicate that the task complete, i.e., `true` or `false`. You will see the effect on the postman response console. To test it:
```
Comment (kayere) [inline articles/a-raw-nodejs-rest-api-without-frameworks-such-as-expressjs/index.md:412]: Fix the typo at the beginning of the sentence.

**Review by kayere — CHANGES_REQUESTED** [review]
Hello @Rose-stack - Great article. Please check my suggestions before we move to tech review. I'd suggest you go through your article before submitting it. I have found a lot of typos which you would have noticed if you reviewed your article. Please go through the whole article so as to catch the typos that I might have missed.

**Comment by Rose-stack** [comment]
Corrected @kayere

**Review by kayere — APPROVED** [review]
Hello @Rose-stack, great work making the changes. The article is now good to go.
@hectorkambow - please give this article a final review.

**Comment by hectorkambow** [comment]
Thank you v much for your patience as we worked through our Repo re-structuring process. 

The purpose behind this restructure - was to better allow us to handle more volume and allow students to publish more articles and get more involved with the community (with open-source contribution and suggestions). 

With most of the changes coming to completion - we will be sure to get your article published asap. 

Thank you again - and we are looking forward to more amazing articles.

**Comment by hectorkambow** [comment]
I'll be running a final review on your PR when ready.
Upon running your article through our 3rd party plagiarism checker it seemed to raise a few flags and the % was higher than we typically accept.
Please see attached PDF - and revisit the article to ensure we are contributing wholly unique and original content.
Be sure to see our resources page to see more info on plagiarism and what is considered as such.


[A Vanilla Node.js REST API Without Frameworks such as Express.js #2243.pdf](https://github.com/section-engi@kayere eering-education/engineering-education/files/6543037/A.Vanilla.Node.js.REST.API.Without.Frameworks.such.as.Express.js.2243.pdf)

@Rose-stack 

cc @kayere

**Comment by Rose-stack** [comment]
The plagiarism stood at 11% on my end 🤔

**Comment by Rose-stack** [comment]
> I'll be running a final review on your PR when ready.
> Upon running your article through our 3rd party plagiarism checker it seemed to raise a few flags and the % was higher than we typically accept.
> Please see attached PDF - and revisit the article to ensure we are contributing wholly unique and original content.
> Be sure to see our resources page to see more info on plagiarism and what is considered as such.
> 
> [A Vanilla Node.js REST API Without Frameworks such as Express.js #2243.pdf](https://github.com/section-engi@kayere eering-education/engineering-education/files/6543037/A.Vanilla.Node.js.REST.API.Without.Frameworks.such.as.Express.js.2243.pdf)
> 
> @Rose-stack
> 
> cc @kayere

@hectorkambow, I can't find the attached PDF, please confirm this

**Comment by hectorkambow** [comment]
[A Vanilla Node.js REST API Without Frameworks such as Express.js #2243.pdf](https://github.com/section-engineering-education/engineering-education/files/6543119/A.Vanilla.Node.js.REST.API.Without.Frameworks.such.as.Express.js.2243.pdf)

**Comment by hectorkambow** [comment]
> The plagiarism stood at 11% on my end 🤔, @hectorkambow

Thank you for checking on your end (that is best practice) - most checkers are different so they may differ or look for different things.

But not sure why it would be so different on the same one? 🤔  

I think many of the red lines may have been code snippets as well. Not sure if those are boilerplate or template code. 

If possible we should change code snippets unless its a needed snippet.

**Comment by Rose-stack** [comment]
Yeah, It could be due to the difference in the checkers  used

**Comment by Rose-stack** [comment]
@hectorkambow I have updated the flags that trigger the higher %. 

I used the same checker as you, But I got a few additional matches pointing to the [review's githubmemory as the source](https://githubmemory.com/@kayere).  Check on your end.
Node js

**Comment by hectorkambow** [comment]
Looks good now - thank you - sorry for the confusion not sure what the issue was.

**Comment by hectorkambow** [comment]
plagiarism check done

**Comment by hectorkambow** [comment]
published https://www.section.io/engineering-education/a-raw-nodejs-rest-api-without-frameworks-such-as-express/

**Pull request closed without merging**